presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewCandidateExtractor.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.common.QualifiedObjectName;
+import com.facebook.presto.sql.tree.DefaultTraversalVisitor;
+import com.facebook.presto.sql.tree.Table;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.facebook.presto.metadata.MetadataUtil.createQualifiedObjectName;
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewCandidateExtractor
+        extends DefaultTraversalVisitor<Void, Void>
+{
+    private final Set<QualifiedObjectName> tableNames = new HashSet<>();
+    private final Map<QualifiedObjectName, List<QualifiedObjectName>> baseTableToMaterializedViews;
+
+    private final Session session;
+
+    public MaterializedViewCandidateExtractor(Session session, Map<QualifiedObjectName, List<QualifiedObjectName>> baseTableToMaterializedViews)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.baseTableToMaterializedViews = requireNonNull(baseTableToMaterializedViews, "baseTableToMaterializedViews is null");
+    }
+
+    @Override
+    protected Void visitTable(Table node, Void context)
+    {
+        tableNames.add(createQualifiedObjectName(session, node, node.getName()));
+        return null;
+    }
+
+    public ImmutableSet<QualifiedObjectName> getMaterializedViewCandidates()
+    {
+        Set<QualifiedObjectName> materializedViewCandidates = new HashSet<>();
```

Comment:
I use Set because the materializedViewCandidates is constantly changing and updating when looping through the table name. Therefore, isEmpty() and retainAll() functions are handy (functions that are absent from ImmutableSet.Builder)